File: app/src/main/java/edu/whu/tmdb/MainActivity.java
package edu.whu.tmdb;

import android.graphics.Typeface;
import android.os.Bundle;
import android.text.Html;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;

import edu.whu.tmdb.R;

public class MainActivity extends AppCompatActivity {

    private EditText etCmd;
    private LinearLayout resultContainer;
    private ScrollView scrollView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        etCmd = findViewById(R.id.etCmd);
        resultContainer = findViewById(R.id.resultContainer);
        scrollView = findViewById(R.id.scrollView);

        getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
        scrollView.setFillViewport(true);

        Button btExecute = findViewById(R.id.btExecute);
        Button btClear = findViewById(R.id.btClear);

        btExecute.setOnClickListener(v -> {
            String sqlCommand = etCmd.getText().toString().trim();
            if (sqlCommand.isEmpty()) {
                Toast.makeText(this, "请输入SQL命令", Toast.LENGTH_SHORT).show();
                return;
            }

            String[] rawResults = Main.execute_UI(sqlCommand);

            for (String rawResult : rawResults) {
                TextView tv = new TextView(this);
                tv.setLayoutParams(new LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.WRAP_CONTENT,
                        LinearLayout.LayoutParams.WRAP_CONTENT));

                String htmlResult = "<pre><tt>" +
                        rawResult.replace(" ", "&nbsp;")
                                .replace("\n", "<br>") +
                        "</tt></pre>";
                tv.setText(Html.fromHtml(htmlResult, Html.FROM_HTML_MODE_LEGACY));

                tv.setTextSize(14);
                tv.setTypeface(Typeface.MONOSPACE);
                tv.setTextColor(0xFF1A1A1A);

                tv.setPadding(4, 4, 4, 4);
                resultContainer.addView(tv);
            }

            scrollView.post(() -> scrollView.fullScroll(View.FOCUS_DOWN));
            etCmd.setText("");
        });

        btClear.setOnClickListener(v -> resultContainer.removeAllViews());
    }
}

File: client/src/main/java/edu/whu/tmdb/client/MainActivity.kt
package edu.whu.tmdb.client

import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.os.RemoteException
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import edu.whu.tmdb.ISqlExecService

class MainActivity : AppCompatActivity() {

    private val targetPackage = "edu.whu.tmdb"
    private val targetAction = "edu.whu.tmdb.action.SQL_EXEC_SERVICE"

    private lateinit var etSql: EditText
    private lateinit var tvStatus: TextView
    private lateinit var tvResult: TextView
    private var remoteService: ISqlExecService? = null
    private var bound = false

    private val connection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            remoteService = ISqlExecService.Stub.asInterface(service)
            bound = true
            tvStatus.text = "Connected: ${name.flattenToShortString()}"
        }

        override fun onServiceDisconnected(name: ComponentName) {
            remoteService = null
            bound = false
            tvStatus.text = "Disconnected"
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        etSql = findViewById(R.id.etSql)
        tvStatus = findViewById(R.id.tvStatus)
        tvResult = findViewById(R.id.tvResult)
        val btnConnect: Button = findViewById(R.id.btnConnect)
        val btnExecute: Button = findViewById(R.id.btnExecute)

        etSql.setText("show ClassTable")

        btnConnect.setOnClickListener { connectService() }
        btnExecute.setOnClickListener { executeSql() }
    }

    private fun connectService() {
        if (bound) {
            tvStatus.text = "Already connected"
            return
        }

        val intent = Intent(targetAction).apply {
            setPackage(targetPackage)
        }
        val ok = bindService(intent, connection, BIND_AUTO_CREATE)
        tvStatus.text = if (ok) "Binding..." else "Bind failed"
    }

    private fun executeSql() {
        val service = remoteService
        if (service == null) {
            tvResult.text = "Service not connected. Tap Connect first."
            return
        }

        val sql = etSql.text?.toString()?.trim().orEmpty()
        if (sql.isEmpty()) {
            tvResult.text = "Please enter SQL."
            return
        }

        try {
            val result = service.executeSql(sql)
            tvResult.text = result.orEmpty()
        } catch (e: RemoteException) {
            tvResult.text = "RemoteException: ${e.message}"
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (bound) {
            unbindService(connection)
            bound = false
        }
        remoteService = null
    }
}
